cleverhans/loss.py

```diff
@@ -0,0 +1,50 @@
+import json
+import os
+import tensorflow as tf
+
+from .model import Model
+
+
+class Loss(object):
+    """
+    An abstract interface for loss wrappers that allows flexible control of
+    real examples, adversarial examples and labels..
+    """
+
+    def __init__(self, model, hparams, attack=None):
+        """
+        :param model: Model instance, the model on which to apply the loss.
+        :param hparams: dict, hyper-parameters for the loss.
+        :param attack: callable, the attack function for adv. training.
+        """
+        assert isinstance(model, Model)
+        assert attack is None or callable(attack)
+        self.model = model
+        self.hparams = hparams
+        self.attack = attack
+
+    def save(self, path):
+        json.dump(dict(loss=self.__class__.__name__,
+                       params=self.hparams),
+                  open(os.path.join(path, 'loss.json'), 'wb'))
+
+    def fprop(self, x, y):
+        """Forward propagate the loss.
+        :param x: tensor, a batch of inputs.
+        :param y: tensor, a batch of outputs (1-hot labels typically).
+        """
+        raise NotImplementedError
+
+
+def attack_softmax_cross_entropy(y, probs, mean=True):
```

Comment:
Why the name `attack_softmax_cross_entropy` and not simply `softmax_cross_entropy`?

Reply:
I renamed because it was confusing as to where this function was used and for what purpose. Basically it's only used by attacks, so I wanted to clearly separate it from defense losses (which may also have softmax crossentropy but follow the Loss API).

Reply:
I think while we're doing the redesign, maybe this would be a good chance to get rid of the "if probs.op.type == 'Softmax'" hack. We've used it for a long time knowing that it wasn't very safe (because sometimes the last op is a reshape, even if there is a softmax right before it).
I think it would be better to standardize on having the attacks call `get_logits` so that we don't need to ask for ancestors in the graph this way.

Reply:
I agree, but it's used quite extensively and there are already a lot of changes.